File: old_trunk/cost/iona/orbacus/java/poa/TestDSIRef_impl.java
// **********************************************************************
//
// Copyright (c) 2000
// Object Oriented Concepts, Inc.
// Billerica, MA, USA
//
// All Rights Reserved
//
// **********************************************************************

package test.poa;

import org.omg.CORBA.*;
import org.omg.PortableServer.*;
import org.omg.PortableServer.POAPackage.*;
import org.omg.PortableServer.ServantLocatorPackage.*;

final class TestDSIRef_impl
    extends org.omg.PortableServer.DynamicImplementation
{
    private ORB orb_;
    private String name_;
    private boolean compare_;
    private boolean defaultServant_;

    //
    // From TestBase (no multiple inheritance)
    //
    public static void
    TEST(boolean expr)
    {
	if(!expr)
	    throw new test.common.TestException();
    }

    TestDSIRef_impl(ORB orb, String name, boolean compare)
    {
	orb_ = orb;
	name_ = name;
	compare_ = compare;
    }

    void
    setDefaultServant(boolean b)
    {
	defaultServant_ = b;
    }

    //
    // Standard IDL to Java Mapping
    //

    public void
    invoke(ServerRequest request)
    {
	String name = request.operation();

	if(!name.equals("aMethod"))
	{
	    NVList list = orb_.create_list(0);
	    request.arguments(list);

	    Any any = orb_.create_any();
	    BAD_OPERATIONHelper.insert(any, new BAD_OPERATION());
	    request.set_exception(any);
	    return;
	}

	//
	// 8.3.1: "Unless it calls set_exception, the DIR must call arguments
	// exactly once, even if the operation signature contains no
	// parameters."
	//
	NVList list = orb_.create_list(0);
	request.arguments(list);

	org.omg.CORBA.Object currentObj = null;
	try
	{
	    currentObj = orb_.resolve_initial_references("POACurrent");
	}
	catch(UserException ex)
	{
	}

	org.omg.PortableServer.Current current = null;
	if(currentObj != null)
	    current = org.omg.PortableServer.CurrentHelper.narrow(currentObj);
	TEST(current != null);

	byte[] oid = null;
	try
	{
	    oid = current.get_object_id();
	}
	catch(org.omg.PortableServer.CurrentPackage.NoContext ex)
	{
	    throw new RuntimeException();
	}
	String oidString = new String(oid);

	if(compare_)
	    TEST(oidString.equals(name_));

//
// Disabled since it is CORBA 2.4
//
/*
        org.omg.PortableServer.Servant servant = null;
        try
        {
            servant = current.get_servant();
        }
        catch(org.omg.PortableServer.CurrentPackage.NoContext ex)
        {
            throw new RuntimeException();
        }
        TEST(servant == this);
*/

	if(defaultServant_)
	{
	    POA poa = null;
	    try
	    {
		poa = current.get_POA();
	    }
	    catch(org.omg.PortableServer.CurrentPackage.NoContext ex)
	    {
		throw new RuntimeException();
	    }
	    byte[] servantId = null;
	    try
	    {
		servantId = poa.servant_to_id(this);
	    }
	    catch(ServantNotActive ex)
	    {
		throw new RuntimeException();
	    }
	    catch(WrongPolicy ex)
	    {
		throw new RuntimeException();
	    }
	    TEST(servantId.length == oid.length);
	    TEST(servantId.equals(oid));
	}
    }

    static final String[] interfaces_ =
    {
        "IDL:Test:1.0"
    };

    public String[]
    _all_interfaces(POA poa, byte[] oid)
    {
        return interfaces_;
    }
}

File: cost/iona/orbacus/cpp/poa/TestDSIRef_impl.cpp
// **********************************************************************
//
// Copyright (c) 2001
// Object Oriented Concepts, Inc.
// Billerica, MA, USA
//
// All Rights Reserved
//
// **********************************************************************

#include <includes.h>

#include <TestDSIRef_impl.h>

using namespace CORBA;
using namespace PortableServer;

TestDSIRef_impl::TestDSIRef_impl(ORB_ptr orb, const char* name,
				 Boolean compare)
    : orb_(ORB::_duplicate(orb)), name_(string_dup(name)), compare_(compare),
      defaultServant_(false)
{
}

TestDSIRef_impl::~TestDSIRef_impl()
{
}

void
TestDSIRef_impl::invoke(ServerRequest_ptr request)
{
    String_var name = request -> operation();

    if(strcmp(name, "aMethod") != 0)
    {
        NVList_ptr list;
        orb_ -> create_list(0, list);
        request -> arguments(list);
    
        Any any;
        any <<= new BAD_OPERATION();
        request -> set_exception(any);
        return;
    }

    //
    // 8.3.1: "Unless it calls set_exception, the DIR must call arguments
    // exactly once, even if the operation signature contains no parameters."
    //
    NVList_ptr list;
    orb_ -> create_list(0, list);
    request -> arguments(list);

    Object_var currentObj;
    try
    {
        currentObj = orb_ -> resolve_initial_references("POACurrent");
    }
    catch(const UserException&)
    {
    }

    PortableServer::Current_var current;
    if(!is_nil(currentObj))
        current = PortableServer::Current::_narrow(currentObj);
    TEST(!is_nil(current));

    ObjectId_var oid = current -> get_object_id();
    String_var oidString = ObjectId_to_string(oid);

    if(compare_)
        TEST(strcmp(oidString, name_) == 0);

#ifndef HAVE_NO_CORBA_2_4
    Servant servant = current -> get_servant();
    TEST(servant == this);
#endif

    if(defaultServant_)
    {
	PortableServer::POA_var poa = current -> get_POA();
	ObjectId_var servantId = poa -> servant_to_id(this);
	TEST(servantId -> length() == oid -> length());
	TEST(memcmp(servantId -> get_buffer(), oid -> get_buffer(),
		    oid -> length()) == 0);
    }
}
    
RepositoryId
TestDSIRef_impl::_primary_interface(const ObjectId& id, POA_ptr poa)
{
    return string_dup("IDL:Test:1.0");
}
